Add the BBODY state transition relation (#884)
IntersectMBO/formal-ledger-specifications: src/Ledger/Conway/Specification/BlockBody.lagda.md

open import Ledger.Conway.Specification.Ledger txs abs
open import Ledger.Conway.Specification.Rewards txs abs
```
-->

# Blocks

The block follows mostly the definition of the Shelley specification. One
notable difference is that the size and the hash of the block body are fields
of the block, instead of functions.

```agda
record BHBody : Type where
  field
    bvkcold : VKey
    bsize   : ℕ
    slot    : Slot
    bhash   : KeyHash
    hBbsize : ℕ

record BHeader : Type where
  field
    bhbody : BHBody
    bhsig  : Sig

record Block : Type where
  field
    bheader     : BHeader
    ts          : List Tx
    bBodySize   : ℕ
    bBodyHash   : KeyHash
    ≡-bBodySize : bBodySize ≡ BHBody.hBbsize (BHeader.bhbody bheader)
    ≡-bBodyHash : bBodyHash ≡ BHBody.bhash (BHeader.bhbody bheader)
```

# Block Body Transition

The Block Body Transition updates the block body state which comprises the
ledger state and the map describing the produced blocks. The environment of the
`BBODY`{.AgdaDatatype} transition are the protocol parameters and the accounting
state.

The helper function `incrBlocks`{.AgdaFunction}, which counts the number of
non-overlay blocks produced by each stake pool.

> The signal is a block from which we extract:

Reviewer: ```suggestion
The signal of the `BBODY`{.AgdaDatatype} rule is a block from which we extract:
```

> * The sequence of transactions txs of the block.
> * The block header body bhb.

Reviewer: ```suggestion
* The sequence of transactions, `txs`{.AgdaBound} = `block`{.AgdaBound} .`ts`{.AgdaField} of the block.
* The block header body `bhb`{.AgdaBound} = `block`{.AgdaBound} .`bheader`{.AgdaField} .`bhbody`{.AgdaField}.
```

> * The verification key vk of the issuer of the block and its hash hk. The
>   transition is executed if the following preconditions are met:

Reviewer: ```suggestion
* The hash of the verification key of the issuer of the block, `hk`{.AgdaBound} = `hash`{.AgdaFunction} (`bhb`{.AgdaBound} .`bvkcold`{.AgdaField}).

The transition is executed if the following preconditions are met:
```

* The size of the block body matches the value given in the block header body.
* The hash of the block body matches the value given in the block header body.

> * The `LEDGERS`{.AgdaBound} transition succeeds.

Reviewer: ```suggestion
* The `LEDGERS`{.AgdaDatatype} transition succeeds.
```
